### PR TITLE
Add Delly somatic SV caller [issue 644 ]

### PR DESCRIPTION
**Pull request by khurrammaqbool**
### This PR:

Added:

- Delly somatic SV caller

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Comment by khurrammaqbool** [comment]
Add exclusion list for delly in cache

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 1]
```diff
@@ -48,3 +48,35 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
 rm -rf {params.tmpdir};
         """
 
+rule delly_tumor_normal:
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:51]: Please follow the order as described in: https://balsamic.readthedocs.io/en/latest/snakemake_etiquette.html

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 2]
```diff
@@ -48,3 +48,35 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+        namemap = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".delly.sample_name_map",
+    params:
+        tmpdir = tempfile.mkdtemp(prefix=tmp_dir),
+        runmode = "local",
+        tumor = get_sample_type(config["samples"], "tumor"),
+        normal = get_sample_type(config["samples"], "normal"),
+        conda = config["bioinfo_tools"].get("delly")
+    singularity: Path(singularity_image, config["bioinfo_tools"].get("delly") + ".sif").as_posix()
+    benchmark:
+        benchmark_dir + 'delly_tumor_normal_' + config["analysis"]["case_id"] + ".delly.tsv"
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:68]:
```suggestion
        benchmark_dir + 'delly_tumor_normal_' + config["analysis"]["case_id"] + "_delly.tsv"
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 3]
```diff
@@ -48,3 +48,35 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+    threads: get_threads(cluster_config, "delly_tumor_normal")
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:69]: Please add a descriptive message after this

File: tests/test_data/references/genome/dly_excl.tsv [hunk 4]
```diff
@@ -0,0 +1,267 @@
+chr1	0	10000	telomere
```
Comment (hassanfa) [inline tests/test_data/references/genome/dly_excl.tsv:1]: You should delete the content of the file, and have it as an empty file. This is a fixture and placeholder for tests.

File: tests/test_data/references/reference.json [hunk 5]
```diff
@@ -13,6 +13,7 @@
         "refGene": "tests/test_data/references/genome/refGene.txt",
         "wgs_calling_interval": "tests/test_data/references/genome/wgs_calling_regions.v1",
         "genome_chrom_size": "tests/test_data/references/genome/hg19.chrom.sizes",
-        "exon_bed": "tests/test_data/references/genome/refseq.flat.bed"
+        "exon_bed": "tests/test_data/references/genome/refseq.flat.bed",
+        "dly_excl": "tests/test_data/references/genome/dly_excl.tsv"
```
Comment (hassanfa) [inline tests/test_data/references/reference.json:17]:
```suggestion
        "delly_exclusion": "tests/test_data/references/genome/delly_exclusion.tsv"
```

File: tests/conftest.py [hunk 6]
```diff
@@ -80,7 +80,9 @@ def reference():
             "rankscore":
             "tests/test_data/references/genome/cancer_rank_model_-v0.1-.ini",
             "access_regions":
-            "tests/test_data/references/genome/access-5k-mappable.hg19.bed"
+            "tests/test_data/references/genome/access-5k-mappable.hg19.bed",
+            "dly_excl":
```
Comment (hassanfa) [inline tests/conftest.py:84]:
```suggestion
            "delly_exclusion":
```

File: tests/conftest.py [hunk 7]
```diff
@@ -80,7 +80,9 @@ def reference():
...
+            "tests/test_data/references/genome/dly_excl.tsv"
```
Comment (hassanfa) [inline tests/conftest.py:85]:
```suggestion
            "tests/test_data/references/genome/delly_exclusion.tsv"
```

**Review by hassanfa — CHANGES_REQUESTED** [review]
Well done!

Some minor comments, a conflict to fix, and also please use longer descriptive names rather than short and abbreviated.

i.e. `dly_excl` --> `delly_exclusion` all over the code.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 3 above)
Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:74]: Added the message

File: tests/test_data/references/genome/dly_excl.tsv (hunk 4 above)
Comment (khurrammaqbool) [reply tests/test_data/references/genome/dly_excl.tsv]: Now the file is empty.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 8]
```diff
@@ -48,3 +52,46 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+        benchmark_dir + 'delly_tumor_normal_' + config["analysis"]["case_id"] + "_delly.tsv"
+    singularity:
+        Path(singularity_image, config["bioinfo_tools"].get("delly") + ".sif").as_posix()
+    params:
+        tmpdir = tempfile.mkdtemp(prefix=tmp_dir),
+        runmode = "local",
+        tumor = get_sample_type(config["samples"], "tumor"),
+        normal = get_sample_type(config["samples"], "normal"),
+        conda = config["bioinfo_tools"].get("delly")
+    threads: get_threads(cluster_config, "delly_tumor_normal")
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:74]:
```suggestion
    threads:
        get_threads(cluster_config, "delly_tumor_normal")
```

Comment (hassanfa) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:74]: new line here as well.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 9]
```diff
@@ -17,16 +17,20 @@ rule manta_tumor_normal:
...
         tmpdir = tempfile.mkdtemp(prefix=tmp_dir),
         runmode = "local",
         tumor = get_sample_type(config["samples"], "tumor"),
         normal = get_sample_type(config["samples"], "normal"),
         conda = config["bioinfo_tools"].get("manta")
-    singularity: Path(singularity_image, config["bioinfo_tools"].get("manta") + ".sif").as_posix() 
-    benchmark: 
-        benchmark_dir + 'manta_tumor_normal_' + config["analysis"]["case_id"] + ".manta.tsv"
-    threads: get_threads(cluster_config, "manta_tumor_normal") 
+    threads: get_threads(cluster_config, "manta_tumor_normal")
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:30]:
```suggestion
    threads:
        get_threads(cluster_config, "manta_tumor_normal")
```

Comment (hassanfa) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:30]: new line

File: CHANGELOG.rst [hunk 10]
```diff
@@ -5,6 +5,11 @@ Added:
 ^^^^^^
 
 * samtools flagstats and stats to workflow and MultiQC
+* Delly v0.8.7 somatic SV caller
+* Delly containter
+* bcf to vcf command
```
Comment (hassanfa) [inline CHANGELOG.rst:10]: We can delete this since it is part of the Delly rule's `shell` already.
```suggestion
```

**Review by hassanfa — APPROVED** [review]
Well done! Some minor changes I made. Merge when tests are done.

File: CHANGELOG.rst [hunk 11]
```diff
@@ -18,7 +22,7 @@ Fixed:
 * vcf-filtered-clinical tag files will have all variants including PASS
 * Refactor snakemake align rules according to snakemake etiquette 
 * Refactor snakemake fastqc vep contest and mosdepth rules according to snakemake etiquette
-
+* Refactor snakemake manta and delly rules according to snakemake etiquette
```
Comment (ashwini06) [inline CHANGELOG.rst:25]: Changelog reflects the changes we made w.r.t develop/master branches. As delly rule is very recent and I think we don't need to add this to changelog

```suggestion
* Refactor snakemake manta rule according to snakemake etiquette
```

**Review by ashwini06 — COMMENTED** [review]
Nice job 👍  Few comments from my end

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 12]
```diff
@@ -48,3 +53,47 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+
+delly call -x {input.excl} -o {params.tmpdir}/delly.bcf -g {input.fa} {input.bamT} {input.bamN}
+
+echo TUMOR$'\t'tumor$'\n'NORMAL$'\t'control > {params.tmpdir}/samples.tsv
+
+delly filter -p -f somatic -o {params.tmpdir}/delly.somatic.bcf -s {params.tmpdir}/samples.tsv {params.tmpdir}/delly.bcf
+
+bcftools view -O z -o {params.tmpdir}/delly.vcf.gz {params.tmpdir}/delly.somatic.bcf
+
+cp {params.tmpdir}/delly.vcf.gz {output.final};
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:92]: Fixed

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 13]
```diff
@@ -48,3 +53,47 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+        """
+source activate {params.conda};
+
+delly call -x {input.excl} -o {params.tmpdir}/delly.bcf -g {input.fa} {input.bamT} {input.bamN}
+
+echo TUMOR$'\t'tumor$'\n'NORMAL$'\t'control > {params.tmpdir}/samples.tsv
+
+delly filter -p -f somatic -o {params.tmpdir}/delly.somatic.bcf -s {params.tmpdir}/samples.tsv {params.tmpdir}/delly.bcf
+
+bcftools view -O z -o {params.tmpdir}/delly.vcf.gz {params.tmpdir}/delly.somatic.bcf
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:90]: Fixed

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 14]
```diff
@@ -48,3 +53,47 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+
+cp {params.tmpdir}/delly.vcf.gz {output.final};
+
+tabix -p vcf -f {output.final};
+
+echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap};
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:96]: {params.tmpdir}/samples.tsv is for delly internal filter command and is different from {output.namemap}.

File: BALSAMIC/utils/constants.py [hunk 15]
```diff
@@ -379,6 +385,15 @@
             "output_file": "access_5kb_hg38.txt",
             "output_path": "genome"
         },
+        "delly_exclusion": {
+            "url":
+            "https://raw.githubusercontent.com/dellytools/delly/master/excludeTemplates/human.hg38.excl.tsv",
+            "file_type": "text",
+            "gzip": False,
+            "genome_version": "hg19",
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/utils/constants.py:393]: Fixed

File: CHANGELOG.rst (hunk 11 above)
Comment (khurrammaqbool) [reply CHANGELOG.rst:29]: fixed

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 14 above)
[the remark this answers is not in the document]
Comment (ashwini06) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:96]: Is that delly accepts Normal samples name as only 'control', whereas in balsamic namemap we have given names for NORMAL samples as 'normal'

**Review by ashwini06 — APPROVED** [review]

File: BALSAMIC/workflows/reference.smk [hunk 16]
```diff
@@ -336,3 +339,15 @@ vep_install --SPECIES {params.species} \
...
+rule prepare_delly_exclusion:
+    input:
+        delly_exclusion = delly_exclusion_url.get_output_file,
+    log:
+        os.path.join(basedir, "genome", "delly_exclusion.log"),
+    singularity:
+        Path(singularity_image, config["bioinfo_tools"].get("delly") + ".sif").as_posix()
+    shell:
+        """
+sed -i 's/chr//g' {input.delly_exclusion} 2> {log}
```
Comment (ashwini06) [inline BALSAMIC/workflows/reference.smk:352]: Is the output a 3-column bed file? Then why not output to '.txt'/ '.bed' file. Any specific reasons to output results to log file?

File: CHANGELOG.rst [hunk 17]
```diff
@@ -5,12 +5,19 @@ Added:
...
+* Delly v0.8.7 somatic SV caller
+* Delly containter
+* bcftools v1.12 to delly container
+* tabix v0.2.6 to delly container
 * PASSed SV calls from Manta to clinical delivery
 * An extra filter to VarDict tumor-normal to remove variants with STATUS=Germline, all other will still be around
 * Added ``vcf2cytosure``` to annotate container
 * ``git`` to the container definition
+* prepare_delly_exclusion rule
+=======
```
Comment (ashwini06) [inline CHANGELOG.rst:17]:
```suggestion
=======
```
This lines indicates the latest additions to changelog (or called conflicts). Resolve them on your local branch and push them.

**Review by ashwini06 — COMMENTED** [review]
@khurrammaqbool : 2 minor comments

File: BALSAMIC/workflows/reference.smk (hunk 16 above)
Comment (khurrammaqbool) [reply BALSAMIC/workflows/reference.smk:352]: The file already has .tsv extension.

Comment (hassanfa) [inline BALSAMIC/workflows/reference.smk:352]: It is not transparent what this rule is doing. It is readable it is replacing the `chr` from a file, but it does not have an output. Looks like a rogue rule. Can you please add an output to it, and then use that output for `rule all`?

**Review by hassanfa — CHANGES_REQUESTED** [review]
`prepare_delly_exclusion` needs an output. Otherwise, looks like a rogue rule.

File: BALSAMIC/workflows/reference.smk (hunk 16 above)
Comment (khurrammaqbool) [reply BALSAMIC/workflows/reference.smk]: @hassanfa The output file is added

**Review by hassanfa — APPROVED** [review]
💯 awesome! Merge ready 😃

**Pull request merged**